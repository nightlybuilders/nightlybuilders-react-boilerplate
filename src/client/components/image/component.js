import React from 'react'
import PropTypes from 'prop-types'
import cx from 'bem-classnames'
import get from 'lodash.get'

// Styling
const classes = {
  name: 'image',
}

export const getImagePath = (src, cachebuster) => {
  // if the src contains http, we assume it is pointing to an external host and
  // return the src right away
  if (src.indexOf('http') >= 0) {
    return src
  }

  return `/__static__/${cachebuster}/images/${src}`
}

// TODO: create ResponsiveImage Version (not working in IE <=11):
//  - https://developer.mozilla.org/en-US/docs/Learn/HTML/Multimedia_and_embedding/Responsive_images)
export const Image = props => (
  <img
    alt={props.alt || ''}
    className={cx(classes, props.className)}
    onClick={props.onClick}
    onKeyDown={props.onClick}
    role="presentation"
    style={props.styles}
    src={getImagePath(props.src, get(props, 'data.cachebuster', 0))}
  />
)

Image.displayName = 'Image'
Image.propTypes = {
  alt: PropTypes.string.isRequired,
  className: PropTypes.string,
  onClick: PropTypes.func,
  src: PropTypes.string.isRequired,
  styles: PropTypes.object, // eslint-disable-line react/forbid-prop-types
}

Image.defaultProps = {
  className: '',
  onClick: () => {},
  styles: {},
}
