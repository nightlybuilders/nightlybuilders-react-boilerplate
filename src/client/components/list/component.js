/* eslint-disable no-underscore-dangle */
import React from 'react'
import PropTypes from 'prop-types'
import BemHelper from 'react-bem-helper'

// Styling
const classes = new BemHelper('list')

export const List = props => {
  const { data, actions } = props
  return (
    <ul {...classes()}>
      {data.length > 0 && data.map(d => <li key={actions.getKey(d)}>{actions.getValue(d)}</li>)}
    </ul>
  )
}

List.displayName = 'List'
List.propTypes = {
  actions: PropTypes.shape({
    getKey: PropTypes.func,
    getValue: PropTypes.func,
  }),
  data: PropTypes.array, // eslint-disable-line react/forbid-prop-types
}
List.defaultProps = {
  actions: {
    getKey: d => d,
    getValue: d => d.id,
  },
  data: [],
}
