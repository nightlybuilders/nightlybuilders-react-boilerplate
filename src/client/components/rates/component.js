/* eslint-disable no-underscore-dangle */
import React from 'react'
import get from 'lodash.get'
import PropTypes from 'prop-types'
import BemHelper from 'react-bem-helper'
import map from 'lodash.map'

import { List, ListItem } from '../list'

// Styling
const classes = new BemHelper('rates')

export const Rates = props => {
  const { data } = props
  return (
    <div {...classes()}>
      <h3>Prefetched Rates (Apollo)</h3>
      <List>
        {map(get(data, 'rates') || [], rate => (
          <ListItem key={rate.currency}>{rate.currency}</ListItem>
        ))}
      </List>
    </div>
  )
}

Rates.displayName = 'Rates'
Rates.propTypes = {
  data: PropTypes.shape({
    rates: PropTypes.array, // eslint-disable-line react/forbid-prop-types
  }),
}
Rates.defaultProps = {
  data: {
    rates: [],
  },
}
