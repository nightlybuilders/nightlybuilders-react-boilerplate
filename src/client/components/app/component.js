/* eslint-disable jsx-a11y/anchor-is-valid */
import React, { Component } from 'react'
import PropTypes from 'prop-types'
import dbg from 'debug'

// Utils
import get from 'lodash.get'
import BemHelper from 'react-bem-helper'
import invoke from 'lodash.invoke'

// Components
import { Button } from '../button'
import { Image } from '../image'
import { List } from '../list'
import { Rates } from '../rates'

// Styling
const classes = new BemHelper('app')

const debug = dbg('nb:App')

export class App extends Component {
  componentDidMount = () => {
    // Illustration: how to refresh data from Apollo
    setTimeout(() => {
      debug('refreshData')
      invoke(this.props, 'data.refetch')
    }, 10000)
  }

  onClick = change => {
    const { actions = {} } = this.props
    invoke(actions, 'onClick', change)
  }

  onNavigate = (path, args) => {
    const { actions = {} } = this.props
    invoke(actions, 'onNavigate', path, args)
  }

  render() {
    const { otherProps, store } = this.props

    return (
      <div {...classes()}>
        <Image {...classes('image')} {...otherProps.image} />
        <h1 {...classes('headline')}>{get(otherProps, 'title')}</h1>
        <p>Current State with Counter: {get(store, 'app.data.counter')}</p>
        <div {...classes('container', 'buttons')}>
          <Button onClick={() => this.onClick(1)}>+1</Button>
          <Button onClick={() => this.onClick(-1)}>-1</Button>
        </div>
        <div {...classes('container')}>
          <Button onClick={() => this.onNavigate('example', { hash: '#some-hash' })}>
            Open Example Page
          </Button>
        </div>
        <div {...classes('container')}>
          <h3>Prefetched Posts (Redux)</h3>
          <List />
        </div>
        <div {...classes('container')}>
          <Rates />
        </div>
      </div>
    )
  }
}

App.propTypes = {
  actions: PropTypes.shape({
    onClick: PropTypes.func,
  }),
  data: PropTypes.object, // eslint-disable-line react/forbid-prop-types
  otherProps: PropTypes.shape({
    image: PropTypes.shape({
      alt: PropTypes.string,
      src: PropTypes.string,
    }),
    title: PropTypes.string,
  }),
  store: PropTypes.shape({
    app: PropTypes.shape({
      data: PropTypes.shape({
        title: PropTypes.string,
      }),
    }),
  }),
}
App.defaultProps = {
  actions: {
    onClick: () => {},
  },
  data: {},
  otherProps: {
    image: {
      alt: '',
      src: '',
    },
    title: 'Hello',
  },
  store: {
    app: {
      data: {
        counter: 0,
      },
    },
  },
}
