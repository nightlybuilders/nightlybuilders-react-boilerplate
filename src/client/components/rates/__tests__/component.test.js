import React from 'react'
import { shallow } from 'enzyme'

import { Rates } from '../component'

describe('Component/Rates', () => {
  const rates = [{ currency: 'EUR' }, { currency: 'USD' }]

  test('should render without throwing an error', () => {
    const wrapper = shallow(<Rates data={{ rates }} />)
    expect(wrapper).toMatchSnapshot()
  })
})
