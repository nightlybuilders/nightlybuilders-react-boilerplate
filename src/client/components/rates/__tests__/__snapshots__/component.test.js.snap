// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component/Rates should render without throwing an error 1`] = `
<div
  className="rates"
>
  <h3>
    Prefetched Rates (Apollo)
  </h3>
  <List
    className=""
  >
    <ListItem
      className=""
      key="EUR"
      onClick={[Function]}
      onKeyPress={[Function]}
      style={Object {}}
    >
      EUR
    </ListItem>
    <ListItem
      className=""
      key="USD"
      onClick={[Function]}
      onKeyPress={[Function]}
      style={Object {}}
    >
      USD
    </ListItem>
  </List>
</div>
`;
